### DISCUSSION TITLE
Confocal scanning saved images and parameters

### DISCUSSION THREAD

**Discussion by qeg-ungar** [Ulm-IQO/qudi-iqo-modules#140] [Q&A] [answered]
I am having issues reloading the most recent scans and scan parameters after closing and restarting qudi, even after saving the scans. It seems the scan data and parameters that repopulate the GUI are not being stored in the same directory where the scan data is being saved. Anyone experience similar issues and/or know of a solution?

**Comment by timoML** [comment]
The scan history is saved as "[Status Variable](https://github.com/Ulm-IQO/qudi-core/blob/main/docs/design_concepts/status_variables.md)". This is completely independent of the data that you save actively. 

If the scan history is not properly recovered after a qudi restart, there should be an error in the logs on module activation.
Do you get such an error or can you describe your unexpected behavior in more detail?

**Reply by qeg-ungar** [reply]
Thanks for the reply. I have attached some screenshots including the log after initializing and screenshots of the scanning interface when first opening (image_open1): 
![image_open1](https://github.com/Ulm-IQO/qudi-iqo-modules/assets/154924563/affd1cea-a20e-4346-bee8-917222252d7e)

updating the scanning parameters and rescanning (image_new_scan):

![image_new_scan](https://github.com/Ulm-IQO/qudi-iqo-modules/assets/154924563/7bfc3819-125f-4546-9c49-e4549c6ef90f)
 
and after quitting qudi and reactivating the scanning module (image_reopen):
![image_reopen](https://github.com/Ulm-IQO/qudi-iqo-modules/assets/154924563/7634da13-cfe5-4bdd-aa1e-8d70bc664f0d)

 As you can see, the scanning parameters return to the settings when originally opened. Log below:
![log_scanning_activcate](https://github.com/Ulm-IQO/qudi-iqo-modules/assets/154924563/8cd8b6a2-2a4f-44a1-bbfd-65148b7b3d42)

**Comment by timoML** [comment]
The scan settings are recovered from the ScanData object. 
In your example, the value for the y axis changes on reloading of the module.
However, the ScanData that is restored here is a "x, z" scan. So it does not contain information on the settings of the y axis, which is expected behavior. Qudi then falls back to some default value.
The correct y scan setting should be displayed when you scroll through the scan history with the "Backward" button.

I understand that this can be a bit confusing. At this time, I do not see an easy way to make things more intuitive, though.


-edit-
Maybe I missed your issue here. Is it that the new scan is not saved to the history at all?

**Reply by qeg-ungar** [reply]
Yes the new scan is not saved to history

**Reply by qeg-ungar** [reply]
additionally, the positions for X,Y,Z were all changed between the original scan (93.18, 144.12, 92.12) and the new scan (95.17, 146.86, 94.85) but then when reopening they returned to the original position (93.18, 144.12, 92.12)

**Reply by timoML** [reply]
Ok, then it's a bug that we haven't seen so far.
I noticed that your qudi version is a bit older. Pulling from main could already help.

For debugging, two things that could go wrong are:
1. The scanned image is never saved to scan history. You could check whether after a scan, scanning_data_logic.get_current_scan_data() is returning the correct scan. If not, looking with a debugger into _update_scan_state() (after the sigScanStateChanged is emitted) might help you understand why this happens.

2. Dumping or loading of the status variables fails.
Please check whether the scan shows up in your AppData folder (on Windows C:/Users/*user_name*/AppData/Roaming/qudi) in status-ScanningDataLogic_logic_scanning_data_logic.cfg.

**Reply by qeg-ungar** [reply]
I tried doing a clean install for qudi-core and qudi-iqo-modules with the most up to date versions, and wiping the AppData contents. Problem still persists. The new scan does not show up in the status-ScanningDataLogic_logic_scanning_data_logic.cfg file.

**Reply by qeg-ungar** [reply]
Update: Qudi forgets the scan history and previous settings if we kill the terminal running Qudi before closing the program "properly" (without a "Shutdown complete! Ciao" message).

What is frustrating is that if we close using the red X, then qudi is still running in the terminal, and needs to be closed with CTRL+C in order to properly save the scan data. 

Thanks for the help

**Reply by timoML** [reply, marked as answer]
The intended way to close qudi is 
a) in the manager: File -> Quit Qudi
b) from the tray icon: Right click -> Quit
If this is not resulting in a "Shutdown complete!" message, your IDE/terminal might not be setup correctly. What are you using there?

Are you saying the scan history is only forgotten if you kill qudi in the terminal? Then, qudi has no time to dump to the disk, and the behavior would be expected. 
If you don't see scans in AppData/status-... on proper shutdown, one would need to dig into the on_deactivate function of the ScanningDataLogic.

**Reply by Neverhorst** [reply]
You can also hit the "Ctrl+Q" shortcut while the qudi manager window is in focus.

Closing any window in qudi does not deactivate the respective module, nor does it close qudi if you close the manager. From the tray icon you can always make each window reappear.
If you think this default behaviour needs changing, please feel free to raise this topic in a separate discussion (or issue).

Regarding the loss of scan history if qudi is not closed properly:
This issue has been discussed before and I would agree that procedural saving after each complete scan would be the ideal solution. However so far no one looked into how to implement that in a good way. Maybe you could also raise that as a [New Feature] issue here so it is on the radar.

**Reply by NicolasStaudenmaier** [reply]
At this point I would like to mention that @TobiasSpohn is planning a save status variables button in the main GUI and will be soon ready to review/merge. See branch [save_status_variables in the qudi-core repository](https://github.com/Ulm-IQO/qudi-core/tree/save_status_variables). An automatic saving every couple of minutes is also planned there but requires some more active testing. 

Saving scan data (status variables) should be easily possible by calling the `_dump_status_variables()` after a scan has finished. This call is a function of `Base` from which every Qudi module inherits.
